### ISSUE TITLE
The `config()` macro should validate the ID it captures from the enclosing `val`

### ISSUE THREAD

**Issue by sjrd** [sbt/librarymanagement#134]
The `config()` macro captures the name of the enclosing `val` to determine the "id" of the configuration. In this snippet:
```scala
val fooBar = config("foo-bar")
```
it will capture the id `fooBar`. The macro does not complain.

Instead, *at run-time* (I mean, build load time), I get the following obscure error:
```
...
[debug]   Load.loadUnit(file:/localhome/doeraene/projects/scalajs/sbt-plugin-test/referencedCrossProject/, ...) took 5034.22581ms

[debug] Load.apply: load took 19346.912803ms

[debug] Load.apply: resolveProjects took 0.660934ms

[debug] Load.apply: finalTransforms took 140.090565ms

[debug] Load.apply: config.delegates took 1.482166ms

[error] java.lang.IllegalArgumentException: requirement failed: id must be capitalized: scalaJSProjectBaseSettings

[error]         at scala.Predef$.require(Predef.scala:277)

[error] java.lang.IllegalArgumentException: requirement failed: id must be capitalized: scalaJSProjectBaseSettings

[error] Use 'last' for the full log.

[debug] > Exec(load-failed, None, None)

[debug] > Exec(last, None, None)

Project loading failed: (r)etry, (q)uit, (l)ast, or (i)gnore? q
```
Note that this is the output *with* `last`. In my case `scalaJSProjectBaseSettings` is the `fooBar`, and it was very obscure because the `config()` call was deep within the rhs of `scalaJSProjectBaseSettings` (which was a `Seq[Setting[_]]`.

The `config()` macro should really validate the id at compile-time. There is no point in delaying this error until run-time, when there is not even a stack trace to figure out what's wrong.

**Comment by dwijnand** [comment]
Thanks for the ticket, @sjrd.

As an aside, I think this also goes for the `project` macro - aren't there runtime checks for the project name too?

**Comment by tomjadams** [comment]
FWIW, I am seeing the same issue after upgrading to sbt 1.0.1, with the New Relic plugin here: https://github.com/gilt/sbt-newrelic/blob/master/src/main/scala/NewRelic.scala#L51

They have an issue to resolve it [1], but it'd be great if they knew about it at compile time, as they've flagged the plugin as being 1.0 ready [2].

[1] https://github.com/gilt/sbt-newrelic/pull/42
[2] https://github.com/sbt/sbt/wiki/sbt-1.x-plugin-migration

**Comment by frgomes** [comment]
I've stumbled with such assertion too.
This is breaking a plugin which I'm trying to upgrade to SBT1.x/Scala2.12.

Question: Any particular reason for the id to be forcefully capitalized?

I find this requirement very inconvenient since it's perfectly natural to employ [triplets](https://wiki.osdev.org/Target_Triplet) or other identifiers which are all lowercase.

Ivy [documentation on configurations](https://ant.apache.org/ivy/history/2.2.0/ivyfile/configurations.html) does not enforce any capitalized names. So, I would expect to able to employ all lowercase IDs in SBT.

**Comment by dwijnand** [comment]
It's making what was a convention (e.g `Compile`, `Test`, etc) a requirement so that the unified slash syntax (which shipped in sbt 1.1) is unambiguous. so `Foo / bar` means "bar key in Foo config" and "foo / bar" means "bar task in foo key".

**Comment by huntc** [comment]
I experienced this problem given:

```
lazy val Cover = project // <-- If the `Cover` is just `cover` then the illegal arg appears
  .in(file("cover"))
  .enablePlugins(SitePlugin)
  .settings(
    SiteScaladocPlugin.scaladocSettings(
      config("core"),
      mappings in(Compile, packageDoc) in core,
      s"api/core"
    )
  )
  .dependsOn(core)
```

In this case, the enclosing `val` is the project. Thus when `config` is inlined as above then the check for an enclosing `val` seems incorrect.

My work-around:

```
lazy val cover = project
  .in(file("cover"))
  .enablePlugins(SitePlugin)
  .settings(
    SiteScaladocPlugin.scaladocSettings(
      {
        val Core = config("core")
        Core
      },
      mappings in(Compile, packageDoc) in streambedCore,
      "api/core"
    )
  )
  .dependsOn(core)
```
